### PR TITLE
PIM-4758: Fixed attribute filtering on method getFullProducts

### LINKED ISSUE #3028: Get full product with only some attributes
I'm a Symfony/Doctrine/Akeneo beginner ;-) and try to get the full data of a product with only some attributes. 

I've seen that `Pim\Bundle\DataGridBundle\Controller\ProductExportController` use for the export this line to get all attributes:

        $results = $this->productRepository->getFullProducts($productsList, $attributeIds);

I've tried the same only with some attribute ids as second parameter of `getFullProducts`, but this doesn't work: I always get all attributes instead of only some of them. (I'm using MySQL.)

I looked down to this line in ProductRepository https://github.com/akeneo/pim-community-dev/blob/master/src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Repository/ProductRepository.php#L250 :

            ->leftJoin('v.attribute', 'a', $qb->expr()->in('a.id', $attributeIds)) 

Shouldn't be the condition the fourth instead of the third parameter here?

### PR DESCRIPTION
**Pull request by ClementGautier**
| Q                  | A
| ------------------ | ---
| Migration script?  | N
| Behats             | N/A
| Specs & Static     | Y
| Changelog updated  | Y
| Fixed issues | #3028

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepository.php [hunk 1]
```diff
@@ -225,17 +225,22 @@ public function getFullProduct($id)
...
             ->addSelect('c, assoc, g')
-            ->leftJoin('v.attribute', 'a', $qb->expr()->in('a.id', $attributeIds))
+            ->leftJoin('v.attribute', 'a')
             ->leftJoin('p.categories', 'c')
             ->leftJoin('p.associations', 'assoc')
             ->leftJoin('p.groups', 'g')
             ->where($qb->expr()->in('p.id', $productIds));
 
+        if (!empty ($attributeIds)) {
+
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepository.php]: blank line to remove

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepository.php [hunk 2]
```diff
@@ -225,17 +225,22 @@ public function getFullProduct($id)
...
         $qb
             ->addSelect('c, assoc, g')
-            ->leftJoin('v.attribute', 'a', $qb->expr()->in('a.id', $attributeIds))
+            ->leftJoin('v.attribute', 'a')
             ->leftJoin('p.categories', 'c')
             ->leftJoin('p.associations', 'assoc')
             ->leftJoin('p.groups', 'g')
             ->where($qb->expr()->in('p.id', $productIds));
 
+        if (!empty ($attributeIds)) {
```
Comment (phaseinducer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepository.php]: `empty()` is not a function so no space before parenthesis imho

File: spec/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepositorySpec.php [hunk 3]
```diff
@@ -18,4 +22,56 @@ function it_is_a_product_repository()
     {
         $this->shouldImplement('Pim\Bundle\CatalogBundle\Repository\ProductRepositoryInterface');
     }
+
+    function it_filters_on_attributes_on_getFullProducts_function($em, QueryBuilder $queryBuilder, Expr $expr, AbstractQuery $query)
```
Comment (phaseinducer) [inline spec/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepositorySpec.php]: not sure it's a good practice to mention a method name here, it's implementation

Comment (ClementGautier) [inline spec/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepositorySpec.php]: @phaseinducer  I didn't find any better name. Do you have one in mind ?

Comment (sumbobyboys) [inline spec/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepositorySpec.php]: and for sure, everything should be lower case :)

Comment (sumbobyboys) [inline spec/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepositorySpec.php]: it_filters_on_attributes_when_get_full_products_function_is_called

File: spec/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepositorySpec.php [hunk 4]
```diff
@@ -18,4 +22,56 @@ function it_is_a_product_repository()
...
+        $em->createQueryBuilder()->willReturn($queryBuilder);
+        $query->execute()->shouldBeCalled();
+
+        $queryBuilder->expr()->willReturn($expr);
+        $queryBuilder->getQuery()->willReturn($query);
+
+        $this->getFullProducts([42], [1, 3, 4]);
+    }
+
+    function it_does_not_filter_with_empty_attributes_on_getFullProducts_function($em, QueryBuilder $queryBuilder, Expr $expr, AbstractQuery $query)
```
Comment (sumbobyboys) [inline spec/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepositorySpec.php]: same here

File: spec/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepositorySpec.php (hunk 3 above)
Comment (phaseinducer) [inline spec/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepositorySpec.php]: no, you still mention implementation
what about `it_joins_only_specified_attributes_when_finding_full_products` ?

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 ## Bug fixes
 - PIM-4706: Product association import error with identifier containing comma or semicolon
+- Issue #3028: Fixed attribute filtering on method getFullProducts
```
Comment (jjanvier) [inline CHANGELOG.md]: PIM-3028

File: spec/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepositorySpec.php (hunk 3 above)
Comment (ClementGautier) [inline spec/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepositorySpec.php]: @phaseinducer Like this one, I keep it :P thx

**Comment by jjanvier** [comment]
ticket number in the title of the PR please

**Comment by jjanvier** [comment]
no new behat ?

File: spec/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepositorySpec.php (hunk 3 above)
Comment (jjanvier) [inline spec/Pim/Bundle/CatalogBundle/Doctrine/ORM/ProductRepositorySpec.php]: +1 yohan

File: CHANGELOG.md (hunk 5 above)
Comment (ClementGautier) [inline CHANGELOG.md]: @jjanvier There is no ticket in Jira for this one, do I need to create one ?

Comment (phaseinducer) [inline CHANGELOG.md]: It's https://github.com/akeneo/pim-community-dev/issues/3028 I think

**Comment by ClementGautier** [comment]
@jjanvier We cant do a behat test on this one because nothing changed: The only feature that use this function is the quick export on 1.3 and the output is the same with or without this patch. btw this function will be deprecated in 1.4 as we don't use it anymore.

File: CHANGELOG.md (hunk 5 above)
Comment (jjanvier) [inline CHANGELOG.md]: ok sorry, then Gitub-3028
don't know if you should create a ticket or not, see with nidup

**Comment by jjanvier** [comment]
ok, just see about the ticket with @nidup then GTM
thanks :)

File: CHANGELOG.md (hunk 5 above)
Comment (nidup) [inline CHANGELOG.md]: yep you can create a ticket in Jira

**Comment by nidup** [comment]
GTM once ticket created and if CI is blue (i currently have issue to run the CI on 1.3)

**Pull request merged**